### DISCUSSION TITLE
Task09 demo_poly.m

### DISCUSSION THREAD

**Discussion by NicolasFranklin** [chulminy/CIVE497-CIVE700#138] [General]
Just out of curiosity, for nn_demo_poly, does the code know that the polynomial is 4th order and it is just fitting the coefficients? Or is it an arbitrary n'th order polynomial which best minimizes the root mean square error?

Thanks

**Comment by chulminy** [comment]
NN is a black box model. We do not assume anything about the model. 

Thus, we are not trying to estimate the coefficient.

**Reply by NicolasFranklin** [reply]
If thats the case, then what determines the order of the polynomial? Is there a way to retrieve the final function that the NN fits to the data?

**Reply by chulminy** [reply]
They are all valid questions!

It's a black box model, so it is not something that you try to find "the order of the polynomial". 
You have a basic NN model and it's a nonlinear complex model that can fit your data (in this case, your polynomial curve). 

The training process is to find the network parameters (w and b). 
There will no visible function that you can explain and analytically evaluate like equations. 
You only have a "model" that gives you the output when you put the input. 

If it can be represented by such analytical equations, they are not a black box model. 

A bright side of NN is to build any complex model regardless of its true complexity. 
A drawback is we do not explain how it works because it's a black box. 

Is that enough?

**Reply by NicolasFranklin** [reply]
Yes, that helps. Thanks!

**Comment by ryankong123** [comment]
To follow up on this question, does that mean that this code technically should work for any function if we give it different data points? ie. if I wanted to generate data for a step function would this code try to fit a polynomial to my step function data or would it just treat it like a black box?

**Reply by chulminy** [reply]
Yes. It should work. 
A goal of NN training is to find w and b which can minimize abs(f(in_train, w,b) - out_train ).
Here, we just design a "structure" or "architecture" of NN (e.g., # of layers, activation function, # of node) depending on the complexity of the problem.
